### PR TITLE
Item actions

### PR DESCRIPTION
**Pull request by Alecaddd**
Created sub-classes of parent `CanvasItem` to handle different shapes.

### CODE REVIEW

File: src/Widgets/CanvasItem.vala [hunk 1]
```diff
@@ -76,6 +83,14 @@ public class GtkCanvas.CanvasItem : Clutter.Actor {
     }
     private double _rotation = 0.0;
 
+    /**
+    * Fill color of the shape
+    */
+    public string color;
```
Comment (Philip-Scott) [inline src/Widgets/CanvasItem.vala]: I think this doesn't belong on the "Abstract" class that all items will use. Each shape might have it's own implementation. And it should also be a a property implementing the get/set methods :)

File: src/Widgets/CanvasItem.vala [hunk 2]
```diff
@@ -136,4 +151,12 @@ public class GtkCanvas.CanvasItem : Clutter.Actor {
 
         updated ();
     }
+
+    internal void apply_rotation (double rotation) {
```
Comment (Philip-Scott) [inline src/Widgets/CanvasItem.vala]: This was made into a property :) so it can be deleted

https://github.com/Philip-Scott/libgtkcanvas/pull/13/files#diff-5f880ccfd30c8b8d210c6ab7dc02f8c3R74

File: src/Widgets/Shapes/ShapeCircle.vala [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+* You should have received a copy of the GNU General Public
+* License along with this program; if not, write to the
+* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+* Boston, MA 02110-1301 USA
+*
+* Authored by: Alessandro Castellani <castellani.ale@gmail.com>
+*/
+
+/**
+ * ShapeCircle is a child class of the parent CanvasItem {@link Clutter.Actor} that handles move, rotate, scale, etc.
```
Comment (Philip-Scott) [inline src/Widgets/Shapes/ShapeCircle.vala]: I don't think you need to mention `that handles move, rotate, scale, etc` :)

File: src/Widgets/Shapes/ShapeCircle.vala [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+public class GtkCanvas.ShapeCircle : GtkCanvas.CanvasItem {
+    public ShapeCircle (string color, double rotation) {
+        this.color = color;
+        this.rotation = rotation;
+
+        real_x = 0;
+        real_y = 0;
+
+        real_w = 100;
+        real_h = 100;
```
Comment (Philip-Scott) [inline src/Widgets/Shapes/ShapeCircle.vala]: Instead of using these values, use https://github.com/Philip-Scott/libgtkcanvas/pull/13/files#diff-5f880ccfd30c8b8d210c6ab7dc02f8c3L109

In the future, the real_n properties will have it's own get/set methods, but if you're going to set all four use the function insteadd to prevent extra calls :)

File: src/Widgets/CanvasItem.vala (hunk 1 above)
Comment (Philip-Scott) [reply src/Widgets/CanvasItem.vala]: Each shape should implement their own `background_color: Clutter.Color.from_string (color)` like so :)

File: src/Widgets/Shapes/ShapeCircle.vala [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        _canvas.set_size (100, 100);
+
+        set_rectangle (null, null, 100, 100);
+
+        _canvas.draw.connect((ctx, w, h) => {
+            ctx.set_source_rgb (0.0, 0.8, 0.0);
+            ctx.arc (w/2, h/2, w/2, angle1, angle2);
+            ctx.fill ();
+
+            // Create the stroke
```
Comment (Philip-Scott) [inline src/Widgets/Shapes/ShapeCircle.vala]: Don't leave commented code ;)

File: src/Widgets/Shapes/ShapeRectangle.vala [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+ * This is a specific shape class to handle the generation of a Rectangular geometry
+ */
+public class GtkCanvas.ShapeRectangle : GtkCanvas.CanvasItem {
+    public ShapeRectangle (string color, double rotation) {
+        this.color = color;
+        this.rotation = rotation;
+
+        background_color = Clutter.Color.from_string (color);
+        
+        real_x = 0;
```
Comment (Philip-Scott) [inline src/Widgets/Shapes/ShapeRectangle.vala]: same as the circle :)

File: src/Widgets/Shapes/ShapeRectangle.vala [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+ * ShapeRectangle is a child class of the parent CanvasItem {@link Clutter.Actor} that handles move, rotate, scale, etc.
+ * 
+ * This is a specific shape class to handle the generation of a Rectangular geometry
+ */
+public class GtkCanvas.ShapeRectangle : GtkCanvas.CanvasItem {
+    public ShapeRectangle (string color, double rotation) {
+        this.color = color;
+        this.rotation = rotation;
+
+        background_color = Clutter.Color.from_string (color);
```
Comment (Philip-Scott) [inline src/Widgets/Shapes/ShapeRectangle.vala]: this is a good place where you could implement color as a property, that sets the background color as well :)

File: src/Utils/HoverEffect.vala [hunk 8]
```diff
@@ -20,65 +20,24 @@
...
-            shadow_size, shadow_spread);
-
-        if (shadow != null) {
-            material.set_layer (0, shadow);
-        }
-
-        var opacity = actor.get_paint_opacity () * shadow_opacity / 255;
-        var alpha = Cogl.Color.from_4ub (255, 255, 255, opacity);
-        alpha.premultiply ();
+        var color = Cogl.Color.from_4ub (65, 201, 253, 255);
```
Comment (Philip-Scott) [inline src/Utils/HoverEffect.vala:38]: Looks like the changes no this file are leaving the "ghost lines" when moving/resizing. Maybe it should be left on a separate PR? :)

File: src/Widgets/Canvas.vala [hunk 9]
```diff
@@ -123,17 +124,22 @@ public class GtkCanvas.Canvas : Gtk.AspectFrame {
         add (stage);
     }
 
-   /**
+    /**
     * Adds a test shape. Great for testing the library!
     *
+    * @param type the shape to be generated [rectangle, circle, triangle]
```
Comment (Philip-Scott) [inline src/Widgets/Canvas.vala]: Since we don't yet have a triangle, don't add a triangle on the docs ;)

**Review by Philip-Scott — CHANGES_REQUESTED** [review]
At the end of the day, i don't think we should have a specific Class for each shape. Once we have a class that handles the drawing of SVG, we can just use that + SVGs of basic shapes :) 

At the end, this library should only have the Base CanvasItem class with the following Items: 

- SVGShapeItem
- TextItem
- ImageItem (For rendering images like PNG)

For now, it's ok to have these classes to be able to test the rest of the work of the code :)

File: src/Widgets/Canvas.vala [hunk 10]
```diff
@@ -33,6 +33,7 @@ public class GtkCanvas.Canvas : Gtk.AspectFrame {
      */
     public signal void item_selected (CanvasItem item, Clutter.ModifierType modifiers);
 
+    public CanvasItem item;
```
Comment (Philip-Scott) [inline src/Widgets/Canvas.vala]: Don't declare item here :) You can declare the item inside of a function without initializing by just doing `CanvasItem item`

File: src/Widgets/Canvas.vala [hunk 11]
```diff
@@ -123,17 +124,22 @@ public class GtkCanvas.Canvas : Gtk.AspectFrame {
...
     * @param color the color the test-shape will be, in CSS format
     * @param rotation the amount of degrees the item will be rotated
     */
-    public CanvasItem add_test_shape (string color, double rotation) {
-        var item = new CanvasItem ();
-        item.background_color = Clutter.Color.from_string (color);
-
-        item.rotation = rotation;
+    public CanvasItem add_shape (string type, string color, double rotation) {
+        switch (type) {
```
Comment (Philip-Scott) [inline src/Widgets/Canvas.vala]: item should be declared here, on top of switch :)

**Review by Philip-Scott — CHANGES_REQUESTED** [review]

File: src/Utils/HoverEffect.vala (hunk 8 above)
Comment (Alecaddd) [reply src/Utils/HoverEffect.vala:38]: This seems to happen only with a border bigger than 1px.
I still feel that having the shadow it's a bit of an overkill. I'd like to merge this to have it this simple hover effect for now. I'll work on a separated branch just for this, to handle it better and actually trying to get the real share of the Cairo object inside the Actor.
What do you think?

File: src/Widgets/Canvas.vala (hunk 11 above)
Comment (Alecaddd) [reply src/Widgets/Canvas.vala]: If I do that, I get an error while compiling.
```
Canvas.vala:145.9-145.24: error: use of possibly unassigned local variable `item'
        add_item (item);
```

Comment (Philip-Scott) [reply src/Widgets/Canvas.vala]: You can always null-check 😉

File: src/Utils/HoverEffect.vala (hunk 8 above)
Comment (Philip-Scott) [reply src/Utils/HoverEffect.vala:38]: If you're going to fix it in another branch you should either make the border 1px or revert it :) just to keep master as working as possible.

It may not really be your fault though 🤔 since the border is "outside" of the actor it might be a drawing bug and we need to force a "re-draw" on that section

Comment (Alecaddd) [reply src/Utils/HoverEffect.vala:38]: Yup, I reverted to 1px! It's the lightest solution so far, and it doesn't seem to leave any ghost lines on the canvas :+1:

File: src/Widgets/Canvas.vala (hunk 11 above)
Comment (Philip-Scott) [reply src/Widgets/Canvas.vala]: Or instead of a null check, you can have a `default:` on the switch :)

Comment (Alecaddd) [reply src/Widgets/Canvas.vala]: :+1:

File: src/Widgets/Shapes/ShapeCircle.vala [hunk 12]
```diff
@@ -35,11 +35,22 @@ public class GtkCanvas.ShapeCircle : GtkCanvas.CanvasItem {
...
     private string _color;
-    private Clutter.Color clutter_color;
+
+    public Clutter.Color? clutter_color {
+        get {
+            return _clutter_color;
+        } set {
+            if (value == null) return;
+
+            clutter_color = value;
```
Comment (Philip-Scott) [inline src/Widgets/Shapes/ShapeCircle.vala]: This asignment to itself is going to infinite loop

File: src/Widgets/Shapes/ShapeCircle.vala [hunk 13]
```diff
@@ -54,7 +65,7 @@ public class GtkCanvas.ShapeCircle : GtkCanvas.CanvasItem {
         set_rectangle (0, 0, 100, 100);
 
         _canvas.draw.connect((ctx, w, h) => {
-            ctx.set_source_rgb (clutter_color.red, clutter_color.green, clutter_color.blue);
+            ctx.set_source_rgb (_clutter_color.red, _clutter_color.green, _clutter_color.blue);
```
Comment (Philip-Scott) [inline src/Widgets/Shapes/ShapeCircle.vala]: you should use the property value instead :) without the `_`

File: src/Widgets/Shapes/ShapeRectangle.vala [hunk 14]
```diff
@@ -51,7 +62,7 @@ public class GtkCanvas.ShapeRectangle : GtkCanvas.CanvasItem {
         set_rectangle (0, 0, 100, 100);
 
         _canvas.draw.connect((ctx, w, h) => {
-            ctx.set_source_rgb (clutter_color.red, clutter_color.green, clutter_color.blue);
+            ctx.set_source_rgb (_clutter_color.red, _clutter_color.green, _clutter_color.blue);
```
Comment (Philip-Scott) [inline src/Widgets/Shapes/ShapeRectangle.vala]: same here

File: src/Widgets/Shapes/ShapeRectangle.vala [hunk 15]
```diff
@@ -35,11 +35,22 @@ public class GtkCanvas.ShapeRectangle : GtkCanvas.CanvasItem {
...
     private string _color;
-    private Clutter.Color clutter_color;
+
+    public Clutter.Color? clutter_color {
+        get {
+            return _clutter_color;
+        } set {
+            if (value == null) return;
+
+            clutter_color = value;
```
Comment (Philip-Scott) [inline src/Widgets/Shapes/ShapeRectangle.vala]: and here :)

File: src/Widgets/Shapes/ShapeRectangle.vala [hunk 16]
```diff
@@ -35,11 +35,22 @@ public class GtkCanvas.ShapeRectangle : GtkCanvas.CanvasItem {
             if (value == "") return;
 
             _color = value;
-            clutter_color = Clutter.Color.from_string (value);
+            _clutter_color = Clutter.Color.from_string (value);
```
Comment (Philip-Scott) [inline src/Widgets/Shapes/ShapeRectangle.vala]: and also set the property instead of the field (`_`), so that you get the nulchecks and all

File: src/Widgets/Shapes/ShapeCircle.vala [hunk 17]
```diff
@@ -35,11 +35,22 @@ public class GtkCanvas.ShapeCircle : GtkCanvas.CanvasItem {
             if (value == "") return;
 
             _color = value;
-            clutter_color = Clutter.Color.from_string (value);
+            _clutter_color = Clutter.Color.from_string (value);
```
Comment (Philip-Scott) [inline src/Widgets/Shapes/ShapeCircle.vala]: Set the property instead of the field here :)

**Review by Philip-Scott — CHANGES_REQUESTED** [review]

File: src/Widgets/Canvas.vala [hunk 18]
```diff
@@ -149,17 +149,27 @@ public class GtkCanvas.Canvas : Gtk.AspectFrame {
...
+
+        switch (type) {
+            case "rectangle":
+                item = new ShapeRectangle (color, rotation);
+            break;
+            case "circle":
+                item = new ShapeCircle (color, rotation);
+            break;
+            default:
+                item = null;
```
Comment (Philip-Scott) [inline src/Widgets/Canvas.vala]: adding a null might break :) Instead maybe use the rectangle

**Review by Philip-Scott — APPROVED** [review]
